### PR TITLE
Add page reloading in offline.php templates

### LINKED ISSUE #438: Automatically reload the offline page once the user is back online
See https://github.com/GoogleChrome/web.dev/pull/4717

### PR DESCRIPTION
**Pull request by thelovekesh**
Fixes #438  

Automatically reload the offline page once the user is back online.

Tasks Done:

- [x] Add JS to automatically reload the offline page once the user is back online

https://user-images.githubusercontent.com/54371619/152957784-e01b9259-ccde-46f3-8f85-4d0404e7b436.mp4

### CODE REVIEW

File: wp-includes/template.php [hunk 1]
```diff
@@ -171,3 +171,47 @@ function wp_service_worker_error_details_template( $output = '' ) {
...
+			window.location.reload();
+		});
+
+		/**
+		 * Check if the server is responding and reload the page if it is.
+		 * This handles the case when the device is online, but the server is offline or misbehaving.
+		 */
+		async function checkNetworkAndReload() {
+			try {
+				const response = await fetch('.', {method: 'HEAD'});
```
Comment (westonruter) [inline wp-includes/template.php]: Any reason for `.` as opposed to `location.href`? The problem with `.` is it will drop the query parameters. 

```suggestion
				const response = await fetch(location.href, {method: 'HEAD'});
```

File: wp-includes/template.php [hunk 2]
```diff
@@ -171,3 +171,47 @@ function wp_service_worker_error_details_template( $output = '' ) {
...
+				// Verify we get a valid response from the server
+				if (response.status >= 200 && response.status < 500) {
+					window.location.reload();
+					return;
+				}
+			} catch {
+				console.log('User is offline');
+			}
+			window.setTimeout(checkNetworkAndReload, 2500);
```
Comment (westonruter) [inline wp-includes/template.php]: Any reason specifically for 2,500ms? I suppose just because this is what was used in https://github.com/GoogleChrome/web.dev/pull/4717.

If the server is having trouble due to heavy traffic or some other reason, getting hit every 2.5 seconds will add additional load and may be interpreted by the host as a DOS attack. Therefore, I suggest the value here implement [exponential backoff](https://en.wikipedia.org/wiki/Exponential_backoff) for retrying. So there could be a counter in the script for the number of tries, and each time the logic runs and the counter is incremented, the next timeout delay can be multiplied by the counter. 

Count | Delay
------|------
1 | 2.5s
2 | 5s
3 | 10s
4 | 20s
5 | 40s

In other words, multiple the 2500 delay by `Math.pow( 2, count )` where the initial count is `0`.

File: wp-includes/template.php [hunk 3]
```diff
@@ -171,3 +171,47 @@ function wp_service_worker_error_details_template( $output = '' ) {
...
+ * Reload the offline page and check if user comes online.
+ *
+ * @since 0.7
+ */
+function wp_service_worker_offline_page_reload() {
+	if ( ! is_offline() ) {
+		return;
+	}
+	?>
+	<script>
```
Comment (westonruter) [inline wp-includes/template.php]: Should this go through JS Coding Guidelines? https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/

Or should we rather apply Prettier formatting?

File: wp-includes/template.php [hunk 4]
```diff
@@ -171,3 +171,47 @@ function wp_service_worker_error_details_template( $output = '' ) {
...
 	echo '<p><!--WP_SERVICE_WORKER_ERROR_MESSAGE--></p>';
 }
+
+/**
+ * Reload the offline page and check if user comes online.
+ *
+ * @since 0.7
+ */
+function wp_service_worker_offline_page_reload() {
+	if ( ! is_offline() ) {
```
Comment (westonruter) [inline wp-includes/template.php]: It would be useful to also do the reload in the case of server being down.

Also, something else that comes to mind: what happens if the error page is served in response to a POST request? If so, each reload will cause a re-POST message from the browser and that's probably not desired.

```suggestion
	if ( ! is_offline() || ! is_500() || 'GET' !== $_SERVER['REQUEST_METHOD'] ) {
```

File: wp-includes/template.php [hunk 5]
```diff
@@ -171,3 +171,47 @@ function wp_service_worker_error_details_template( $output = '' ) {
...
+				if (response.status >= 200 && response.status < 500) {
+					window.location.reload();
+					return;
+				}
+			} catch {
+				console.log('User is offline');
+			}
+			window.setTimeout(checkNetworkAndReload, 2500);
+		}
+		checkNetworkAndReload();
```
Comment (westonruter) [inline wp-includes/template.php:219]: Either the entire script should be wrapped in an IIFE, or the `<script>` should have `type="module"` added to it. This will prevent the `checkNetworkAndReload` function from polluting the global scope.

**Review by westonruter — CHANGES_REQUESTED** [review]

File: wp-includes/template.php (hunk 3 above)
Comment (thelovekesh) [reply wp-includes/template.php]: > apply Prettier formatting

~~Sounds good to me.~~

I think JS code in PHP files that follows WPCS should follow WP JS standards.

File: wp-includes/template.php (hunk 1 above)
Comment (thelovekesh) [reply wp-includes/template.php:200]: That's interesting. Above we are making a fetch request to base origin to check whether we are getting response OK or not. I wonder how making a fetch request to some origin can drop query params of another origin.

File: wp-includes/template.php (hunk 2 above)
Comment (thelovekesh) [reply wp-includes/template.php]: I think we should also reset the counter after the count is quals to 5 or 6. There can be a possibility that the user or server comes back online and page reloading has to wait for another 1 or 2 minutes due to time delay. What do you think?

Comment (ankitrox) [reply wp-includes/template.php]: @thelovekesh We can cap the number of retries, so after `n` number of retries, there won't be more.

File: wp-includes/template.php (hunk 4 above)
Comment (thelovekesh) [reply wp-includes/template.php]: Thanks for pointing this out. For sure we need to check the method of the request. I have implemented these changes.

File: wp-includes/template.php (hunk 3 above)
Comment (westonruter) [reply wp-includes/template.php]: Note that core may be moving to Prettier: https://make.wordpress.org/core/2022/01/05/proposal-changes-to-javascript-coding-standards-for-full-prettier-compatibility/

**Comment by westonruter** [comment]
Note that testing this is a bit tricky because the script is part of the precached template, so without a version bump it won't show up.

**Comment by westonruter** [comment]
Let's just add a PHPUnit test to ensure that the cases are covered for this function outputting the `script`.

File: wp-includes/template.php (hunk 2 above)
Comment (thelovekesh) [reply wp-includes/template.php]: @westonruter Please let me know your thoughts on it.

Comment (westonruter) [reply wp-includes/template.php]: I think the exponential backoff will be enough. After 15 tries the timeout goes up to ~1 day, so it will effectively be disabled once the count goes high enough.

**Comment by thelovekesh** [comment]
> Note that testing this is a bit tricky because the script is part of the precached template, so without a version bump it won't show up.

@westonruter  Tests added 👍

File: wp-includes/template.php [hunk 6]
```diff
@@ -171,3 +171,57 @@ function wp_service_worker_error_details_template( $output = '' ) {
...
+				}
+			} catch {
+				// Unable to connect so do nothing.
+			}
+			window.setTimeout(checkNetworkAndReload, Math.pow(2, count++) * 2500);
+		}
+		checkNetworkAndReload();
+	</script>
+	<?php
+	return ob_get_clean();
```
Comment (westonruter) [inline wp-includes/template.php]: Why adding the return here? If I'm not mistaken, the output buffering is causing the `script` to no longer be printed onto the page? The output buffering should rather be done in the tests.

Comment (thelovekesh) [reply wp-includes/template.php]: > output buffering is causing the script to no longer be printed

@westonruter Ahh. my bad. Might get confused during the testing. 
Fixed this now 👍

**Comment by thelovekesh** [comment]
@westonruter Can you please confirm if this PR needs no more improvements?

**Review by westonruter — APPROVED** [review]
This is good, yes.

**Comment by westonruter** [comment]
I just realized one thing that should be done to improve this. Namely, if you try going to `/?wp_error_template=offline` then the result is an infinite reload as it tries to go back online. Accessing this URL directly is key for being able to see how the offline template looks during development. So I think the logic to auto-reload should be disabled if this is true:

```js
( new URLSearchParams(location.search.substr(1)) ).has( 'wp_error_template' )
```

**Comment by thelovekesh** [comment]
> then the result is an infinite reload 

Thanks for pointing it out and it's a great finding. I just checked it after your comment. We should add it else it will cause infinite reloads during development. 

@westonruter Should I push a commit for this?

**Comment by westonruter** [comment]
@thelovekesh Yes please! Thanks.

File: wp-includes/template.php (hunk 4 above)
Comment (westonruter) [reply wp-includes/template.php]: As part of #728, I just realized that this check will never do anything:

https://github.com/GoogleChromeLabs/pwa-wp/blob/124baabe30f70b718960fa18158f10a54943287e/wp-includes/template.php#L181-L184

This is because the offline template is always fetched via a `GET` request by the service worker when it is installing, so it will never not be `GET`. We need an alternative way of preventing the code from running on responses to non-GET requests.

Contrary to what I thought above, each reload does _not_ cause the browser re-POST message: https://github.com/GoogleChromeLabs/pwa-wp/pull/728#issuecomment-1099718487

Comment (westonruter) [reply wp-includes/template.php]: I believe I've addressed this via https://github.com/GoogleChromeLabs/pwa-wp/pull/763

Comment (westonruter) [reply wp-includes/template.php]: See https://github.com/GoogleChromeLabs/pwa-wp/pull/763
